docs/codeql/codeql-language-guides/analyzing-data-flow-in-go.rst

```diff
@@ -0,0 +1,369 @@
+
+.. code-block:: ql
+
+   import go
+
+   module MyFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+   module MyFlow = DataFlow::Global<MyFlowConfiguration>;
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where data may flow from.
+-  ``isSink`` - defines where data may flow to.
+-  ``isBarrier`` - optional, restricts the data flow.
+-  ``isAdditionalFlowStep`` - optional, adds additional flow steps.
+
+The data flow analysis is performed using the predicate ``flow(DataFlow::Node source, DataFlow::Node sink)``:
+
+.. code-block:: ql
+
+   from DataFlow::Node source, DataFlow::Node sink
+   where MyFlow::flow(source, sink)
+   select source, "Data flow to $@.", sink, sink.toString()
+
+Using global taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Global taint tracking is to global data flow what local taint tracking is to local data flow. That is, global taint tracking extends global data flow with additional non-value-preserving steps. The global taint tracking library is used by applying the module ``TaintTracking::Global<ConfigSig>`` to your configuration instead of ``DataFlow::Global<ConfigSig>``:
+
+.. code-block:: ql
+
+   import go
+
+   module MyFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+   module MyFlow = TaintTracking::Global<MyFlowConfiguration>;
+
+The resulting module has an identical signature to the one obtained from ``DataFlow::Global<ConfigSig>``.
+
+Flow sources
+~~~~~~~~~~~~
+
+The data flow library contains some predefined flow sources. The class ``RemoteFlowSource`` (defined in ``semmle.code.java.dataflow.FlowSources``) represents data flow sources that may be controlled by a remote user, which is useful for finding security problems.
+
+Examples
+~~~~~~~~
+
+This query shows a taint-tracking configuration that uses remote user input as data sources.
+
+.. code-block:: ql
+
+   import go
+
+   module MyFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       source instanceof RemoteFlowSource
+     }
+
+     ...
+   }
+
+   module MyTaintFlow = TaintTracking::Global<MyFlowConfiguration>;
+
+Exercises
+~~~~~~~~~
+
+Exercise 2: Write a query that finds all hard-coded strings used to create a ``url.URL``, using global data flow. (`Answer <#exercise-2>`__)
+
+Exercise 3: Write a class that represents flow sources from ``os.Getenv(..)``. (`Answer <#exercise-3>`__)
+
+Exercise 4: Using the answers from 2 and 3, write a query which finds all global data flows from ``os.Getenv`` to ``url.URL``. (`Answer <#exercise-4>`__)
```

Comment:
"Data flows" sounds odd to me. How about "data flow paths" instead?
```suggestion
Exercise 4: Using the answers from 2 and 3, write a query which finds all global data flow paths from ``os.Getenv`` to ``url.URL``. (`Answer <#exercise-4>`__)
```

Reply:
I've done this for all the language guides where it appears.